pkg/http/handler/timeout_test.go

```diff
@@ -138,53 +186,133 @@ func TestTimeToFirstByteTimeoutHandler(t *testing.T) {
+		idleTimeout:      longIdleTimeout,
+		firstByteTimeout: longFirstByteTimeout,
+		handler: func(*sync.Mutex, chan error) http.Handler {
+			return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				panic(http.ErrAbortHandler)
+			})
+		},
+		wantStatus: http.StatusGatewayTimeout,
+		wantBody:   "request timeout",
+		wantPanic:  true,
+	}, {
+		name:             "timeout before panic",
+		idleTimeout:      shortIdleTimeout,
+		firstByteTimeout: longFirstByteTimeout,
+		handler: func(*sync.Mutex, chan error) http.Handler {
+			return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				time.Sleep(1 * time.Second)
+				panic(http.ErrAbortHandler)
+			})
+		},
+		timeoutMessage: "request timeout",
+		wantStatus:     http.StatusGatewayTimeout,
+		wantBody:       "request timeout",
+		wantPanic:      false,
+	}, {
+		name:             "successful writes prevent timeout",
+		idleTimeout:      shortIdleTimeout,
+		firstByteTimeout: longFirstByteTimeout,
+		handler: func(*sync.Mutex, chan error) http.Handler {
+			return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				w.Write([]byte(""))
+				time.Sleep(shortIdleTimeout * 2 / 3)
+				w.Write([]byte(""))
+				time.Sleep(shortIdleTimeout * 2 / 3)
+				panic(http.ErrAbortHandler)
+			})
+		},
+		wantStatus: http.StatusGatewayTimeout,
+		wantBody:   "request timeout",
+		wantPanic:  true,
+	}, {
+		name:             "can still timeout after a few successful writes",
+		idleTimeout:      shortIdleTimeout,
+		firstByteTimeout: longFirstByteTimeout,
+		handler: func(*sync.Mutex, chan error) http.Handler {
+			return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				w.Write([]byte(""))
+				time.Sleep(shortIdleTimeout * 2 / 3)
+				w.Write([]byte(""))
+				time.Sleep(shortIdleTimeout * 2 / 3)
+				w.Write([]byte(""))
+				time.Sleep(shortIdleTimeout * 2)
+				panic(http.ErrAbortHandler)
+			})
+		},
+		timeoutMessage: "request timeout",
+		wantStatus:     http.StatusOK,
+		wantBody:       "request timeout",
+		wantPanic:      false,
+	}, {
+		name:             "no idle timeout",
+		idleTimeout:      noIdleTimeout,
+		firstByteTimeout: longFirstByteTimeout,
+		handler: func(*sync.Mutex, chan error) http.Handler {
+			return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				panic(http.ErrAbortHandler)
+			})
+		},
+		wantStatus: http.StatusGatewayTimeout,
+		wantBody:   "request timeout",
+		wantPanic:  true,
+	},
+	}
 
-			if rr.Body.String() != test.wantBody {
-				t.Errorf("Handler returned unexpected body: got %q want %q", rr.Body.String(), test.wantBody)
-			}
+	testTimeoutScenario(t, scenarios)
 
-			if test.wantWriteError {
-				if err := <-writeErrors; !errors.Is(err, http.ErrHandlerTimeout) {
-					t.Error("Expected a timeout error, got", err)
-				}
-			}
-		})
-	}
 }
 
-func BenchmarkTimeToFirstByteTimeoutHandler(b *testing.B) {
+func BenchmarkTimeoutHandler(b *testing.B) {
```

Comment:
Let's put the results of this benchmark before/after this change (using benchcmp) in the PR description

Reply:
Added